### PR TITLE
Create Auth API

### LINKED ISSUE #504: Separate User.js into two different classes
Move the following to a separate file:

- register
- login
- setEmailToVerified
- verify

### PR DESCRIPTION
**Pull request by law-nguyen**
Resolves #504 
I hate node.

### CODE REVIEW

File: api/routes/Auth.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  // }
+
+  const decoded = checkIfTokenValid(req);
+  if (!decoded) {
+    res.sendStatus(UNAUTHORIZED);
+  } else {
+    res.status(OK).send(decoded);
+  }
+});
+
```
Comment (hsorathia) [inline api/routes/Auth.js:182]: what do this do

File: api/routes/Auth.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  const newDate = new Date();
+  // + 1 to account for daylight savings time
+  newDate.setDate(newDate.getDate() + 1);
+  const amountOfDaysToLastSunday = newDate.getDate() - newDate.getDay();
+  const lastSundayDate = new Date();
+  lastSundayDate.setDate(amountOfDaysToLastSunday); // last sunday
+  lastSundayDate.setHours(23, 59, 59); // 11:59:59 PM
+
+  // If the last login is before last Sunday
+  // at 1 second before midnight, return true
```
Comment (hsorathia) [inline api/routes/Auth.js:195]: do we need these type of comments

File: src/APIFunctions/User.js [hunk 3]
```diff
@@ -11,7 +11,7 @@ export async function getAllUsers(token) {
   let status = new UserApiResponse();
   await axios
     // get all user!
-    .post('/api/user/users', {
+    .post('/api/User/users', {
```
Comment (hsorathia) [inline src/APIFunctions/User.js:14]: god bless

File: api/mailer/auth.js [hunk 4]
```diff
@@ -9,93 +19,152 @@ const bcrypt = require('bcryptjs');
...
-        });
+        hashedId = hash;
+        res.status(OK).send({ hashedID });
       });
     });
   });
-}
+});
+
+// /* Abstract: Returns a hashed ID based on the
```
Comment (evanugarte) [inline api/mailer/auth.js:51]: pls remove commented out lines

File: api/mailer/auth.js [hunk 5]
```diff
@@ -9,93 +19,152 @@ const bcrypt = require('bcryptjs');
...
-      });
     });
   });
-}
+});
+
+// async function validateVerificationEmail(email, hashedId) {
+//   // Find the user in the DB
+//   return new Promise(async (resolve, reject) => {
+//     await User.findOne({ email: email }, async function(error, result) {
```
Comment (evanugarte) [inline api/mailer/auth.js:134]: remove commented out lines

File: api/mailer/mailer.js [hunk 6]
```diff
@@ -1,75 +1,106 @@
...
-    .catch(_ => {
-      res.sendStatus(NOT_FOUND);
-    });
-});
+// const express = require('express');
+// const router = express.Router();
+// const { SceGoogleApiHandler } = require('./SceGoogleApiHandler');
+// const verification = require('./templates/verification');
+// const { validateVerificationEmail, generateHashedId } = require('./auth');
+// const {
```
Comment (evanugarte) [inline api/mailer/mailer.js:37]: remove commented out lines

File: src/APIFunctions/User.js (hunk 3 above)
Comment (evanugarte) [reply src/APIFunctions/User.js:14]: thanks

File: test/Auth.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        email: 'a@b.c',
+        password: 'password'
+      };
+      const result = await test.sendPostRequest(
+        '/api/Auth/login', user);
+      expect(result).to.have.status(UNAUTHORIZED);
+    });
+
+    /**
+     * failing because user's default acesslevel is NOW -1
```
Comment (evanugarte) [inline test/Auth.js:141]: remove this

File: test/user.js [hunk 8]
```diff
@@ -65,183 +68,49 @@ describe('Users', () => {
...
-        name: 'name',
-        email: 'email',
-        accessLevel: 'accessLevel'
-      });
-      const result = await test.sendPostRequest(
-        '/api/user/verify', { token: token });
-      expect(result).to.have.status(OK);
-    });
-  });
+  // describe('/POST checkIfUserExists with a user added', () => {
```
Comment (evanugarte) [inline test/user.js:95]: remove this if it is done

File: api/mailer/Util.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
+const axios = require('axios');
```
Comment (evanugarte) [inline api/mailer/Util.js:1]: the functions in here have to be a part of the auth API. mailer will be running in a container that doesnt have access to mognodb so this wouldnt work, see below

image

Comment (law-nguyen) [reply api/mailer/Util.js:1]: So I should move this into `api/routes/Auth.js`?

File: api/mailer/Util.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+//   await axios
+//     .post('api/mailer/generateHashedId')
+//     .then(res =>{
+//       hashedID = res;
+//     })
+//     .catch(err => {
+//      hashedID = err;
+//     });
+//   return hashedID;
+// }
```
Comment (hsorathia) [inline api/mailer/Util.js:64]: Are we going to re-include this commented out function later. If not, good idea to remove it xd

File: api/routes/Auth.js (hunk 1 above)
Comment (hsorathia) [reply api/routes/Auth.js:174]: just noting this down bc we have comments on some of the router.post() calls, and then there are none so it feels a bit inconsistent -- we should come back to this when we do documentation

File: test/User.js [hunk 11]
```diff
@@ -73,167 +76,33 @@ describe('User', () => {
...
+  //   it('Should return statusCode 409 when a user +
+  // already exists', async () => {
+  //     const user = {
+  //       email: 'a@b.c'
+  //     };
+  //     const result = await test.sendPostRequest(
+  //       '/api/User/checkIfUserExists', user);
+  //     expect(result).to.have.status(CONFLICT);
+  //   });
+  // });
```
Comment (hsorathia) [inline test/User.js:105]: delete?

File: api/mailer/Util.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+async function generateHashedId(email){
+  let hashedId = '';
+  const toHash = { email : email };
+  await axios
+    .post('http://localhost:8080/api/Auth/generateHashedId', toHash)
+    .then(res =>{
+      hashedId = res.data.hashedId;
+    })
+    .catch(err => {
+      hashedId = err;
```
Comment (evanugarte) [inline api/mailer/Util.js:26]: maybe make it null?
```suggestion
      hashedId = null;
```

File: api/mailer/Util.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+      status = err.data;
+    });
+  return status;
+}
+
+async function generateHashedId(email){
+  let hashedId = '';
+  const toHash = { email : email };
+  await axios
+    .post('http://localhost:8080/api/Auth/generateHashedId', toHash)
```
Comment (evanugarte) [inline api/mailer/Util.js:21]: import this at the top
```suggestion
    .post(`${GENERAL_API_URL}`/api/Auth/generateHashedId', toHash)
```

File: api/mailer/Util.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    })
+    .catch(err => {
+      status = err.data;
+    });
+  return status;
+}
+
+async function generateHashedId(email){
+  let hashedId = '';
+  const toHash = { email : email };
```
Comment (evanugarte) [inline api/mailer/Util.js:19]: remove this and on the .`post` line do

```js
.post('http://localhost:8080/api/Auth/generateHashedId', { email })
```

File: api/mailer/mailer.js [hunk 15]
```diff
@@ -2,14 +2,14 @@ const express = require('express');
...
 const verification = require('./templates/verification');
-const { validateVerificationEmail, generateHashedId } = require('./auth');
 const {
   OK,
   BAD_REQUEST
 } = require('../constants').STATUS_CODES;
 const config = require('../config/config');
 const { USER } = config.googleApiKeys;
 
+
```
Comment (evanugarte) [inline api/mailer/mailer.js:12]: remove

File: api/routes/Auth.js (hunk 1 above)
Comment (evanugarte) [reply api/routes/Auth.js:174]: we need docs on api stuff, you're right

File: api/routes/Auth.js (hunk 2 above)
Comment (evanugarte) [reply api/routes/Auth.js:187]: he pretty much copied and pasted this from `User.js`

File: src/APIFunctions/Auth.js [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+      status.error = true;
+      status.responseData = error.response;
+    });
+  return status;
+}
+
+/**
+ * Checks if the user is signed in by evaluating a jwt token in local storage.
+ * @returns {UserApiResponse} Containing information for
+ *                            whether the user is signed
```
Comment (evanugarte) [inline src/APIFunctions/Auth.js:74]:
```suggestion
 * whether the user is signed or not
```

File: src/APIFunctions/Auth.js [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+ * in or not
```
Comment (evanugarte) [inline src/APIFunctions/Auth.js:75]:
```suggestion
```

File: src/APIFunctions/mailer.js [hunk 18]
```diff
@@ -23,23 +23,4 @@ export async function sendVerificationEmail(email, firstName) {
...
-      email,
-      hashedId
-    })
-    .catch(err => {
-      status.responseData = err;
-      status.error = true;
-    });
-  return status;
-}
+
```
Comment (evanugarte) [inline src/APIFunctions/mailer.js:26]: remove

File: test/User.js (hunk 11 above)
Comment (evanugarte) [reply test/User.js:105]: +1

File: src/APIFunctions/Auth.js [hunk 19]
```diff
@@ -0,0 +1,122 @@
+import axios from 'axios';
+import { UserApiResponse, ApiResponse } from './ApiResponses';
+import { updateLastLoginDate } from './User';
+
+/**
+ * Add a new user to the database.
+ * @param {Object} userToRegister - The object containing all of the user's
+ * information
+ * @param {(string|undefined)} userToRegister.firstName
```
Comment (evanugarte) [inline src/APIFunctions/Auth.js:9]: why would this be undefined? wouldnt it be string?

**Pull request merged**